src/experimental/tools/download2go/index.ts

```diff
@@ -0,0 +1,415 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { IDBPDatabase } from "idb";
+import { AsyncSubject } from "rxjs";
+
+import EventEmitter from "../../../utils/event_emitter";
+import PPromise from "../../../utils/promise";
+
+import { IActiveDownload, IActivePauses } from "./api/context/types";
+import { setUpDb } from "./api/db/dbSetUp";
+import DownloadManager from "./api/downloader/downloadManager";
+import {
+  getBuilderFormattedForAdaptations,
+  getBuilderFormattedForSegments,
+  getKeySystemsSessionsOffline,
+  offlineManifestLoader,
+} from "./api/downloader/manifest";
+import { IContentProtection } from "./api/drm/types";
+import {
+  IContentLoader,
+  IDownload2GoEvents,
+  IEmitterTrigger,
+  IGlobalSettings,
+  IInitSettings,
+  IStoredManifest,
+} from "./types";
+import {
+  checkForPauseAMovie,
+  checkForResumeAPausedMovie,
+  checkInitDownloaderOptions,
+  IndexDBError,
+  isSupported,
+  SegmentConstuctionError,
+  ValidationArgsError,
+} from "./utils";
+
+/**
+ * Instanciate a D2G downloader.
+ * @param {Object<{nameDB, storeManifestEvery}>} IOptionsStarter
+ * @return {IPublicAPI} IPublicAPI
+ */
+class D2G extends EventEmitter<IDownload2GoEvents> {
+  static isPersistentLicenseSupported(): Promise<boolean> {
+    return isSupported();
+  }
+
+  public readonly nameDB: string;
+  private db: IDBPDatabase | null;
+  private emitter: IEmitterTrigger<IDownload2GoEvents>;
+  private activeDownloads: IActiveDownload;
+  private activePauses: IActivePauses;
+
+  constructor(options: IGlobalSettings = {}) {
+    super();
+    this.nameDB =
+      options.nameDB == null || options.nameDB === ""
+        ? "d2g-canalplus"
+        : options.nameDB;
+    this.activeDownloads = {};
+    this.activePauses = {};
+    this.db = null;
+    this.emitter = {
+      trigger: (eventName, payload) => this.trigger(eventName, payload),
+    };
+  }
+
+  /**
+   * Initialize an IndexDB instance.
+   * @returns {Promise<void>}
+   */
+  initDB(): Promise<IDBPDatabase> {
+    return new Promise((resolve, reject) => {
+      setUpDb(this.nameDB)
+        .then(db => {
+          this.db = db;
+          resolve(db);
+        })
+        .catch((error: Error) => {
+          this.trigger("error", { action: "initDB", error });
+          reject(error);
+        });
+    });
+  }
+
+  /**
+   * Start a download from scratch.
+   * @param {Object<ISettingsDownloader>} settings
+   * @returns {Promise.<void>}
+   */
+  async download(options: IInitSettings): Promise<void> {
+    try {
+      const db = this.db;
+      if (db === null) {
+        throw new Error("The IndexDB database has not been created!");
+      }
+      await checkInitDownloaderOptions(options, db, this.activeDownloads);
```

Comment:
So if the contentID is already set, the user has to know by seeing that download returns a failed promise but for most other cases (like when `initDB` fails) you have to listen to an error event?

That's not really consistent.

Reply:
Right, we are now creating the content id for the user and we are handling the error case thanks to a try/catch